### PR TITLE
Nonlinear interp

### PR DESCRIPTION
**Pull request by Rlamboll**
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Example added (either to an existing notebook or as a new notebook, where applicable)
- [x] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <https://github.com/GranthamImperial/silicone/pull/XX>`_) Added feature which does something
- (`#XX <https://github.com/GranthamImperial/silicone/pull/XX>`_) Fixed bug identified in (`#YY <https://github.com/GranthamImperial/silicone/issues/YY>`_)
```

### CODE REVIEW

File: src/silicone/utils.py [hunk 1]
```diff
@@ -181,12 +181,21 @@ def _remove_t0_from_wide_db(times_needed, _db):
             _db.loc[model, scenario, time] = _db.loc[model, scenario, time] - offset
 
 
-def _make_interpolator(variable_follower, variable_leader, wide_db, time_col):
+def _make_interpolator(
```
Comment (znicholls) [inline src/silicone/utils.py:192]: should this be in `utils` or just a private function in `interpolation`?

File: src/silicone/utils.py [hunk 2]
```diff
@@ -181,12 +181,21 @@ def _remove_t0_from_wide_db(times_needed, _db):
...
+    variable_follower, variable_leader, wide_db, time_col, interpkind="linear"
+):
     """
     Constructs a linear interpolator for variable_follower as a function of
     (one) variable_leader for each timestep in the data.
     """
     derived_relationships = {}
+    # We want a generic function to return constants where required
+    def f_factory(ys):
```
Comment (znicholls) [inline src/silicone/utils.py]: Move to a private function? I can't see why this function is nested but maybe I'm missing something

File: src/silicone/utils.py [hunk 3]
```diff
@@ -207,12 +216,26 @@ def _make_interpolator(variable_follower, variable_leader, wide_db, time_col):
...
+                    xs,
+                    ys,
+                    bounds_error=False,
+                    fill_value=(ys[0], ys[-1]),
+                    assume_sorted=True,
+                    kind=interpkind,
+                )
+            else:
+                # We must add boundaries to the spline to prevent extrapolating larger values
+                xs = np.append(np.append(xs[:1] - 1.0, xs), xs[-1:] + 1.0)
```
Comment (znicholls) [inline src/silicone/utils.py]: np.concatenate might remove the need for two calls to append?

File: CHANGELOG.rst [hunk 4]
```diff
@@ -4,10 +4,14 @@ Changelog
...
 master
 ------
+Added
+~~~~~
+- (`#146 <https://github.com/GranthamImperial/silicone/pull/146>`_) Added the ability to do non-linear interpolation.
 
 Changed
 ~~~~~~~
 - (`#142 <https://github.com/GranthamImperial/silicone/pull/142>`_) Sped up RMS closest
+- (`#146 <https://github.com/GranthamImperial/silicone/pull/146>`_) Deprecated the linear interpolator in favour of a generic interpolator.
```
Comment (znicholls) [inline CHANGELOG.rst]:
```suggestion
- (`#146 <https://github.com/GranthamImperial/silicone/pull/146>`_) Deprecated the linear interpolator :class:`LinearInterpolation` in favour of a generic interpolator :class:`Interpolation`.
```

File: CHANGELOG.rst [hunk 5]
```diff
@@ -4,10 +4,14 @@ Changelog
 
 master
 ------
+Added
+~~~~~
+- (`#146 <https://github.com/GranthamImperial/silicone/pull/146>`_) Added the ability to do non-linear interpolation.
```
Comment (znicholls) [inline CHANGELOG.rst]:
```suggestion
- (`#146 <https://github.com/GranthamImperial/silicone/pull/146>`_) Added the ability to do non-linear interpolation by introducing :class:`Interpolation`.
```

**Review by znicholls — APPROVED** [review]
looks good, a few comments but nothing major

File: src/silicone/utils.py (hunk 1 above)
Comment (Rlamboll) [reply src/silicone/utils.py:192]: it's a useful function on its own, I use it in a few notebooks. Arguably it shouldn't begin with the _ in this case

Comment (znicholls) [reply src/silicone/utils.py:192]: Ok nice. If it's useful then making it public is also good (particularly given you've already got the tests, so maybe all it needs is documentation..)

Comment (znicholls) [reply src/silicone/utils.py:192]: Ok nice. If it's useful then making it public is also good (particularly given you've already got the tests, so maybe all it needs is documentation..)

Comment (Rlamboll) [reply src/silicone/utils.py:192]: I've added some documentation for it, I haven't renamed it as that would break the existing uses.

**Pull request merged**